### ISSUE TITLE
*** Assertion failure in -[XCTestContext addUIInterruptionMonitorWithDescription:handler:]

### ISSUE THREAD

**Issue by slozo** [Ahmed-Ali/Cucumberish#47]
There's a problem with adding `UIInterruptionMonitor` in iOS UITests using Swift.

Steps created in Cucumberish do not have `XCTestContextScope` and this causes an assertion.

Here's what I mean:
Before running `addUIInterruptionMonitor(withDescription: "") { ...}` I have this in my `XCTestCase`

> myappCucumberTests.MyappCommonUITestBase	
|>XCFit.XCFit	XCFit.XCFit	
 -|>XCTest.XCTestCase	XCTestCase	
  --|>baseXCTest@0	XCTest	
   ---|>_internalImplementation	_XCTestCaseImplementation *	0x6080003609c0	
    ----|>_testContext	XCTestContext *	0x608000422460
     -----|>_currentScope	id	0x0

Note: `_currentScope` is `nil`
This causes the assertion:
>  Currently executing: "And I add interruption"
 *** Assertion failure in -[XCTestContext addUIInterruptionMonitorWithDescription:handler:], /Library/Caches/com.apple.xbs/Sources/XCTest_Sim/XCTest-11753/XCTestFramework/Classes/XCTestContext.m:113

When I manually set `XCTestContextScope` everything works as expected.
Maybe I'm doing something wrong? 

Xcode Version 8.2.1 (8C1002)

**Comment by slozo** [comment]
OK, I have to run `addUIInterruptionMonitor` on original `XCTestCase` from `implementation.additionalContent`

**Comment by onereddogmedia** [comment]
Hey, you mean something like the following:
`            let xctest = userinfo?["XCTestCase"] as! XCTestCase
            let handler = xctest.addUIInterruptionMonitor(withDescription: "Permission Dialog") { (alert) -> Bool in
                if (alert.buttons["Allow"].exists) {
                    alert.buttons["Allow"].tap()
                } else if (alert.buttons["OK"].exists) {
                    alert.buttons["OK"].tap()
                }
                return true
            }
            app.tap()
            xctest.removeUIInterruptionMonitor(handler)
`

**Comment by slozo** [comment]
That's right. 
However you can get original XCTestCase with NSString constant called `kXCTestCaseKey` instead of writting `"XCTestCase"` as a dictionary key.

**Comment by georgiosantoniadis** [comment]
Even though the issue is several months old, it seems that it is still not really solved intuitively...
How do you get the original `XCTestCase` in, e.g. the `beforeStart` or the `before` block?
I am currently stuck because I cannot add an UIInterruptMonitor for a permission dialog...

**Comment by slozo** [comment]
@georgiosantoniadis can you move the InterruptMonitor into step like: `Given("the app is running")` and make it as a starting point for all your scenarios? I've done that in the past and it worked. I have "WelcomeScreen" with `Given the app is running` step and later on, I'm building on top of that. For example my Login screen has the step: "Given I'm on a login screen" and it calls "Given the app is running" with: `SStep(dict?[kXCTestCaseKey], "Given the app is running")`. 
I don't know if that clear to you ;)

**Comment by georgiosantoniadis** [comment]
I will try that. Thanks a lot for the quick response!

**Comment by rafaelclaycon** [comment]
How do I get `userinfo` in `let xctest = userinfo?["XCTestCase"] as! XCTestCase`? Or in case I decide to use `SStep(dict?[kXCTestCaseKey], "Given the app is running")` where do i get the `dict` from and how to I use this declaration? Does it return something and why does it mention the step definition ("Given the app is running")?

<img width="579" alt="Screen Shot 2019-05-02 at 16 02 43" src="https://user-images.githubusercontent.com/5100081/57099838-c0dbe500-6cf3-11e9-840e-4360584900b3.png">

**Comment by slozo** [comment]
@rafaelclaycon You get the dict as a second parameter in steps. For example something like this:
```
class OnboardingScreenSteps: OnboardingScreen {    
    override func steps() {
 (...)
    Then("Onboarding process is finished and I should see welcome screen") { (_, dict) in
            guard let testCase = dict?[kXCTestCaseKey] as? XCTestCase else {
                XCTFail("No original test case passed to accept notification settings")
                return
            }
            
            self.acceptNotificationSettings(with: testCase)
            SStep(nil, "I should see welcome screen")
        }
    func acceptNotificationSettings(with testCase: XCTestCase) {
        testCase.addUIInterruptionMonitor(withDescription: "NotificationSettings") { (alert) -> Bool in
            alert.buttons["Allow"].tap()
            return true
        }
    }
}
```

**Comment by rafaelclaycon** [comment]
Thanks to your quick and helpful response, I was able to make it work! I would like to emphasize the need to tap or swipe the app (`app.tap()` or `app.swipeUp()`) in order for the interaction with the prompt to go through. In my case I needed to have both an 'Allow' and a 'Don't Allow' scenario for interacting with the camera prompt. Took me a while to figure out the missing piece was the tap or swipe ;). Thanks!
